Apps/Src/SendCarCAN.c

```diff
@@ -1,36 +1,187 @@
- */
+ * @brief print SendCarCAN_Q put/get for debug purposes
+*/
+void print_SendCarCAN_Q(void) {
+    if(SendCarCAN_Q_is_full(&CANFifo)) {
+        printf("\n\rSendCarCAN_Q is full.");
+    }
+    printf("\n\rSendCarCAN_Q put: %d\n\rSendCarCAN_Q get: %d", CANFifo.put, CANFifo.get);
+}
+
+/**
+ * @brief Wrapper to put new message in the CAN queue
+*/
+void SendCarCAN_Put(CANDATA_t message){
+    OS_ERR err;
+    CPU_TS ticks;
+
+    // Reduce the frequency at which we forward motor messages to CarCAN
+    uint8_t msgIdx = message.ID - MOTOR_MSG_BASE_ADDRESS;
+
+    if (msgIdx >= 0 && msgIdx < NUM_MOTOR_MSGS) {  // Check if the message is from the motor controller
+        if (++motorMsgCount[msgIdx] < motorMsgThreshold[msgIdx]) {
+            return;  
+        } else {
+            motorMsgCount[msgIdx] = 0; // Reset counter and continue sending
+        }     
+    }
+    
+    OSMutexPend(&CarCAN_Mtx, 0, OS_OPT_PEND_BLOCKING, &ticks, &err);
+    assertOSError(err);
+    
+    bool success = SendCarCAN_Q_put(&CANFifo, message);
+
+    OSMutexPost(&CarCAN_Mtx, OS_OPT_POST_NONE, &err);
+    assertOSError(err);
+
+    if(success) OSSemPost(&CarCAN_Sem4, OS_OPT_POST_1, &err);
+    assertOSError(err);
+}
+
+/**
+ * @brief Initialize SendCarCAN
+*/
+void SendCarCAN_Init(void) {
+    OS_ERR err;
+    
+    OSMutexCreate(&CarCAN_Mtx, "CarCAN_Mtx", &err);
+    assertOSError(err);
+    
+    OSSemCreate(&CarCAN_Sem4, "CarCAN_Sem4", 0, &err);
+    assertOSError(err);
+
+    SendCarCAN_Q_renew(&CANFifo);
+}
+
+/**
+ * @brief Grabs the latest messages from the queue and sends over CarCAN
+*/
 void Task_SendCarCAN(void *p_arg){
-    CANDATA_t motorMsg;
     OS_ERR err;
+    CPU_TS ticks;
+
+    CANDATA_t message;
+    memset(&message, 0, sizeof(CANDATA_t));
+
+    // PutIOState
+    OSTaskCreate(
+        (OS_TCB*)&putIOState_TCB,
+        (CPU_CHAR*)"PutIOState",
+        (OS_TASK_PTR)Task_PutIOState,
+        (void*)NULL,
+        (OS_PRIO)TASK_SEND_CAR_CAN_PRIO,
+        (CPU_STK*)putIOState_Stk,
+        (CPU_STK_SIZE)WATERMARK_STACK_LIMIT,
+        (CPU_STK_SIZE)TASK_SEND_CAR_CAN_STACK_SIZE,
+        (OS_MSG_QTY)0,
+        (OS_TICK)0,
+        (void*)NULL,
+        (OS_OPT)(OS_OPT_TASK_STK_CLR),
+        (OS_ERR*)&err
+    );
+    assertOSError(err);
+
+    while (1) {
+          
+        // Check if there's something to send in the queue (either IOState or Car state from sendTritium)
+        OSSemPend(&CarCAN_Sem4, 0, OS_OPT_PEND_BLOCKING, &ticks, &err);
+        assertOSError(err);
+
+        OSMutexPend(&CarCAN_Mtx, 0, OS_OPT_PEND_BLOCKING, &ticks, &err);
+        assertOSError(err);
+    
+        bool res = SendCarCAN_Q_get(&CANFifo, &message);
+        assertOSError(err);
 
+        OSMutexPost(&CarCAN_Mtx, OS_OPT_POST_NONE, &err);
+        assertOSError(err);
+
+        if(res) CANbus_Send(message, true, CARCAN);
+        
+    }
+}
+
+static void putIOState(void){
+    CANDATA_t message;
+    memset(&message, 0, sizeof(message));
```

Comment:
Same here, no need for parentheses for `sizeof` if it's used on a variable name.